File: PCC2/lz77.cpp
#include <iostream>
#include <vector>
#include <math.h>
#include <algorithm> 

using namespace std;

class lz77 {
public:
	lz77() {

	}

	string code_int(int n, vector<char> alphabet, int nbits) {
		string code = "";
		int r = alphabet.size();

		int bit = 0;
		while(n) {
			bit = n % r;
			code = alphabet[bit] + code;
			n = n/r;
		}

		string first = "";
		for(int i = 0; i < nbits - code.length(); i++) {
			first += alphabet[0];
			// cout << "len(code): " << nbits - code.length() << endl;
		}

		return first + code;
	}

	int decode_int(string code, vector<char> alphabet, int nbits) {
		int n = 0;
		int r = alphabet.size();
		int rpow = 1;
		int p = code.length() - 1;
		char bit;

		while(p >= 0) {
			bit = code[p];
			n = n + (find(alphabet.begin(), alphabet.end(), bit) - alphabet.begin()) * rpow;
			p = p - 1;
			rpow = rpow * r;
		}

		return n;
	}

	vector<int> prefix_match(string w, int ls) {
		int lmax = 0;
		int p = 0;
		int m = w.length();
		int l;

		for(int s = 0; s < ls; s++) {
			l = 0;
			while((ls + l < m) && (w[s + l] == w[ls + l])) {
				l += 1;
			}

			if(l > lmax) {
				lmax = l;
				p = s;
			}

		}

		vector<int> pl;
		pl.push_back(p);
		pl.push_back(lmax);

		return pl;
	}

	string encode(string txt, int ls, int ll, vector<char> alphabet) {
		txt = txt + "$";
		string code = "";
		int r = alphabet.size();
		int pbits = ceil(log(ls) / log(r));
		int lbits = ceil(log(ll) / log(r));
		int n = txt.length();
		string w = "";

		for(int i = 0; i < ls; i++) {
			w = w + "_";
		}

		w = w + txt.substr(0, min(n, ll));
		int j = -1;

		vector<int> pl;
		char c;

		while(w.length() > ls) {
			// printf("window: %s.%s\n", w.substr(0, ls).c_str(), w.substr(ls).c_str());
			pl = prefix_match(w, ls);
			// printf("p= %d l= %d\n", pl[0], pl[1]);

			c = txt[j + pl[1] + 1];
			code = code + code_int(pl[0], alphabet, pbits) + code_int(pl[1], alphabet, lbits) + c;
			// printf("%d\n", min(n, j + ll + pl[1] + 2));
			if(j+ll+1 > txt.length()) {
				w = w.substr(pl[1]+1);	
			} else {
				w = w.substr(pl[1]+1) + txt.substr(j + ll + 1, (min(n, j + ll + pl[1] + 2) - (j + ll + 1)));
			}
			// cout << "para J= " << j << " W= " << w << endl;
			j += (pl[1] + 1);
		}

		return code;
	}

	string decode(string code, int ls, int ll, vector<char> alphabet) {
		int r = alphabet.size();
		int pbits = ceil(log(ls) / log(r));
		int lbits = ceil(log(ll) / log(r));
		int i = 0;
		int n = code.length();
		
		string txt = "";
		for(int i = 0; i < ls; i++) {
			txt += "_";
		}

		int j = 0;
		int p;
		int l;
		char c;
		int k;

		while(i < n) {
			p = decode_int(code.substr(i, pbits), alphabet, pbits);
			i += pbits;
			l = decode_int(code.substr(i, lbits), alphabet, lbits);
			i += lbits;
			c = code[i];
			i += 1;
			k = 0;

			while(k < l) {
				txt += txt[j + p + k];
				k += 1;
			}

			txt += c;
			j += (l + 1);
		}

		return txt.substr(ls);
	}
};
/*
int main() {
	vector<char> alphabet;
	alphabet.push_back('a');
	alphabet.push_back('b');
	alphabet.push_back('c');
	alphabet.push_back('d');
	alphabet.push_back('r');

	string abr = "abracadabra";
	string txt = "";
	for(int i = 0; i < 20; i++) {
		txt = txt + abr;
	}

	lz77 lz;
	string code = lz.encode(txt, 16, 16, alphabet);

	printf("code= %s\n", code.c_str());

	string decoded = lz.decode(code, 16, 16, alphabet);

	printf("decoded= %s\n", decoded.c_str());

	return 0;
}
*/

File: lz77.py
import math

def code_int(n, alphabet, nbits):
	code = ""
	r = len(alphabet)

	while(n):
		bit = n%r
		code = alphabet[bit] + code
		n = n/r
	code = (nbits - len(code)) * alphabet[0] + code
	
	return code

def decode_int(code, alphabet, nbits):
	n = 0
	r = len(alphabet)
	rpow = 1
	p = len(code) - 1
	while(p >= 0):
		bit = code[p]
		n += (alphabet.index(bit)*rpow)
		p -= 1
		rpow *= r
	return n

def prefix_match(W, Ls):
	lmax = 0
	p = 0
	m = len(W)
	for s in range(Ls):
		l = 0
		while(Ls + l < m and W[s+l] == W[Ls + l]):
			l += 1
		if(l > lmax):
			lmax = l
			p = s
	return (p, lmax)

def lz77_encode(txt, Ls, Ll, alphabet):
	txt = txt + '$'
	code = ""
	r = len(alphabet)
	pbits = int(math.ceil(math.log(Ls, r)))
	lbits = int(math.ceil(math.log(Ll, r)))
	n = len(txt)
	W = Ls * "_" + txt[0:min(n, Ll)]
	
	J = -1
	while(len(W) > Ls):
		# print "window", W[0:Ls]+"."+W[Ls:]
		(p, l) = prefix_match(W, Ls)
		# print "p=", p, "l=", l
		c = txt[J+l+1]
		code += code_int(p, alphabet, pbits) + code_int(l, alphabet, lbits) + c
		W = W[l+1:] + txt[J+Ll+1:min(n, J+Ll+l+2)]
		J += (l+1)
	return code

def lz77_decode(code, Ls, Ll, alphabet):
	r = len(alphabet)
	pbits = int(math.ceil(math.log(Ls, r)))
	lbits = int(math.ceil(math.log(Ll, r)))
	i = 0
	n = len(code)
	txt = (Ls*"_")
	J = 0

	while(i < n):
		p = decode_int(code[i:i+pbits], alphabet, pbits)
		i += pbits
		l = decode_int(code[i:i+lbits], alphabet, lbits)
		i += lbits
		c = code[i]
		i += 1
		# print "p, l, c", p, l, c
		k = 0
		while(k < l):
			txt = txt + txt[J+p+k]
			k += 1
		txt += c
		J += (l+1)
	return txt[Ls:]

def main():
	txt = "abracadabra"
	# txt ="aababbabbac"
	txt = 20*txt
	alphabet = list('abcdr')
	# print txt
	code = lz77_encode(txt, 16, 16, alphabet)

	print "code=", code

	decode = lz77_decode(code, 16, 16, alphabet)

	print "decoded=", decode

	# a = code_int(24, alphabet, 2)

	# print a + "\n"

	# b = decode_int(a, alphabet, 2)

	# print b

if __name__ == '__main__':
	main()